### DISCUSSION TITLE
PSR-12 §6.1: Unary operators

### DISCUSSION THREAD

**Discussion by wimski** [PHP-CS-Fixer/PHP-CS-Fixer#7552] [RFC]
Let me start by quoting the opening line of the `README.md` (emphasis is mine):
> The PHP Coding Standards Fixer (PHP CS Fixer) tool fixes your code _**to follow standards**_

This would let one to assume that when applying a ruleset from a specific standard, the code would only be fixed according to that standard. This is no longer the case for spaces surrounding unary operators in PSR-12 since [this PR](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7388) was merged.

The PSR-12 standard says the following about spaces surrounding unary operators:

[6. Operators](https://www.php-fig.org/psr/psr-12/#6-operators)
> When space is permitted around an operator, multiple spaces **MAY** be used for readability purposes.
> All operators not described here are left undefined.

[6.1 Unary operators](https://www.php-fig.org/psr/psr-12/#61-unary-operators)
> The increment/decrement operators **MUST NOT** have any space between the operator and operand.
> Type casting operators **MUST NOT** have any space within the parentheses.

If the tool would follow the standard for the PSR-12 ruleset, you would expect increment/decrement and type casting operators to be fixed by a _'unary operator spaces'_ rule, but nothing else. Any further changes, which are not defined in the standard, should be up to configuration by the user.

Perhaps the rule could be made configurable in this kind of fashion:
```php
// Apply to all
[
    'unary_operator_spaces' => true, 
];

// Apply to none
[
    'unary_operator_spaces' => false,
];

// Apply to selection (PSR-12 default example)
[
    'unary_operator_spaces' => [
        'increment',
        'decrement',
        'type_casting_parentheses',
    ],
];
```
---

Related sources:
* https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7388
* https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/7303
* https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/7504

**Comment by keradus** [comment]
We need config for the rule then. would you like to raise PR for it, @wimski ?

**Comment by wimski** [comment]
@keradus I can certainly give this a go when I find the time. And I will first have to familiarize myself with this code-base and how configuration is done for comparable rules.

What would be the defaults for the various rulesets?

**Reply by Wirone** [reply]
> What would be the defaults for the various rulesets?

IMHO `unary_operator_spaces` should enable only what's included in PSR/PER for their respective rulesets, and enable everything for `PhpCsFixer` ruleset (to achieve what's configured now). I don't know how to handle `Symfony` ruleset, but I would go with all operators too (so `PhpCsFixer` would inherit it) - @nicolas-grekas can you confirm?

**Reply by keradus** [reply]
I can confirm Sf codebase follows it for all operators

**Comment by keradus** [comment]
Closed by merge of https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7626
